ts/src/profiler.ts

```diff
@@ -13,47 +13,299 @@
+   * If there is a problem polling profiler server for instructions
+   * on the type of profile created, this problem will be logged. If the problem
+   * indicates one definitely will not be able to profile, an error will be
+   * thrown.
+   */
   async start(): Promise<void> {
-    return Promise.reject(
-        new Error('start() is unimplemented for ProfileAgent.'));
+    return this.pollProfilerService();
+  }
+
+  /**
+   * Endlessly polls the profiler server for instructions, and collects and
+   * uploads profiles as requested.
+   */
+  async pollProfilerService(): Promise<void> {
+    const startCreateMillis = Date.now();
+    const prof = await this.createProfile();
+    await this.profileAndUpload(prof);
+    const endCreateMillis = Date.now();
+
+    // Schedule the next profile.
+    setTimeout(this.pollProfilerService.bind(this), 0).unref();
+  }
+
+  /**
+   * Talks to profiler server, which hangs until server indicates
+   * job should be profiled.
+   *
+   * If any problem is encountered, the problem will be logged and
+   * createProfile() will be retried.
+   *
+   * TODO: implement backoff and retry when error encountered. createProfile()
+   * should be retried at time response indicates this request should be retried
+   * or with exponential backoff (up to one hour) if the response does not
+   * indicate when to retry this request. Once this is implemented, an error
+   * will be thrown only if the error indicates one definitely should not
+   * retry createProfile.
+   *
+   * TODO (issue #28): right now, this call could hang for up to an hour when
+   * this method is the only thing on the event loop, keeping the program open
+   * even when all work is done. Should expose the ability to cancel the http
+   * request made here, and then determine when to cancel this request.
+   *
+   * Public to allow for testing.
+   */
+  async createProfile(): Promise<RequestProfile> {
+    const reqBody = {
+      deployment: this.deployment,
+      profileType: this.profileTypes,
+    };
+    const options = {
+      method: 'POST',
+      uri: '/profiles',
+      body: reqBody,
+      json: true,
+    };
+
+    try {
+      const results = await this.request(options);
+      // TODO: check types, don't cast.
+      const body = results[0] as RequestProfile;
+      const response = results[1] as http.ServerResponse;
+
+      if (isErrorResponseStatusCode(response.statusCode)) {
+        this.logger.error('Error creating profile: ' + response.statusMessage);
+      } else {
+        return body;
+      }
+    } catch (err) {
+      this.logger.error('Error creating profile: ' + err.toString());
+    }
+
+    // TODO: determine which codes and errors one should not retry on.
+    // TODO: check response to see if response specifies a backoff.
+    // TODO: implement exponential backoff.
+    await delay(this.config.backoffMillis);
+    return this.createProfile();
```

Comment:
Recursive call. This will stack overflow eventually. Also note that you are creating a chain of promises that never resolve.